[MRG] fix inconsistent random state assignment between parallel trials

rythorpe opened the pull request:
fixes #136 

Running the code [here](https://gist.github.com/rythorpe/e7bac070bff41f108996b98c8b58ac14) creates this plot with `MPI_Backend`:
![mn_mpi_10trials](https://user-images.githubusercontent.com/20212206/92980315-a8411a80-f463-11ea-926f-54720dd8d25e.png)

and this plot for `JoblibBackend`:

![mn_joblib_10trials](https://user-images.githubusercontent.com/20212206/92980331-c3138f00-f463-11ea-997b-624657c21b3b.png)

Note that while individual trials are not exactly the same between the two versions (most likely due to an inconsistency in the trajectories of random states), both show 10 distinct trials that emerged from distinct random states. If anyone can readily spot where the discrepancy is occurring it be much appreciated. Otherwise we might want to address that in another PR since the tests are still passing.

This PR also addresses some of the inconsistent messaging that gets output during a simulation in the terminal.

jasmainak commented on hnn_core/parallel_backends.py:
```diff
@@ -86,10 +86,11 @@ def _clone_and_simulate(self, net, trial_idx):
+            net.params[param_key] += trial_idx
 
         neuron_net = NetworkBuilder(net)
+        neuron_net.net.trial_idx = trial_idx
```
is this equivalent to https://github.com/jonescompneurolab/hnn-core/pull/171/files#diff-ca5f146e027d9dc6ef29ff4db9c21ca2R68 ? I have a suspicion it is not. Can you check?

rythorpe replied:
You're right, I messed this up. Fixed in https://github.com/jonescompneurolab/hnn-core/pull/171/commits/b340fe0150c7b333d40a56f3e87ea0c0ea85004f.

jasmainak commented on hnn_core/network_builder.py:
```diff
@@ -353,6 +352,11 @@ def _create_cells_and_feeds(self):
         External inputs are not targets.
         """
         params = self.net.params
+        # Re-create external feed param dictionaries
```
should we add an "XXX" in the comment somewhere? To indicate that this is a hack (right?) and we should get rid of at some point

rythorpe replied:
Hmm, I'm not sure in this scenario what delineates a hack from a long-term bug fix. We plan to refactor the params and feed modules wrt how feed objects and their respective parameters are passed to `Network` and `NetworkBuilder`. Along the way we will end up simplifying the code by e.g. removing the need for calling `create_pext()` here. That said, I'm happy to update the comment if you think it'd be clearer.

jasmainak replied:
yeah, I agree it might go away but just leaving an XXX tells future contributors that it's a hack that needs to fixed.

jasmainak commented:
I have a hunch that digging [into this](https://github.com/jonescompneurolab/hnn-core/pull/171#discussion_r487367214) might solve your issue. Then ideally, we can add a tiny test to check that MPI vs Joblibs gives same result for 2 trials (with a 3 x 3 network to make it run fast). If it doesn't work, I am fine merging the PR as is if it's good by @blakecaldwell 

do make sure to update `whats_new.rst`. I see that you found a fun way to end your week ;-)

rythorpe commented:
Update with https://github.com/jonescompneurolab/hnn-core/pull/171/commits/b340fe0150c7b333d40a56f3e87ea0c0ea85004f: the script that generated the plots shown at the [beginning of this PR](https://github.com/jonescompneurolab/hnn-core/pull/171#issue-485560556) now produces the same results for the joblib and mpi backends when running the default param set: 

![mn_mpi_10trials_02](https://user-images.githubusercontent.com/20212206/93145424-44645f00-f6ba-11ea-93d3-26c3d8e0fa5c.png)
![mn_joblib_10trials_02](https://user-images.githubusercontent.com/20212206/93145433-4d553080-f6ba-11ea-9dbc-730fa488fe6b.png)

jasmainak commented:
Amazing! do you think we could add a tiny test?

- perhaps one test that checks that the two trials are different
- and another that checks that the results are consistent between joblib and MPI?

you can use a small network, e.g., `params.update({"N_pyr_x": 3, "N_pyr_y": 3})` and maybe a shorter `tstop` to make it run fast?

also don't forget to update `whats_new.rst`!

rythorpe commented:
I've updated `whats_new.rst` and added your suggested tests @jasmainak. I also ended up rearranging some of the test functions in `test_compare_hnn.py` in https://github.com/jonescompneurolab/hnn-core/pull/171/commits/b8f5d1d5935cc756c478b5ef2bae92e1cec17508 in order to limit redundant code. Take a look and let me know if the changes are clear and acceptable.

jasmainak commented:
Fantastic, thanks for making the changes. PR looks good to me, merged!
